src/server.c

```diff
@@ -3466,8 +3466,14 @@ int processCommand(client *c) {
      * Note that we do not want to reclaim memory if we are here re-entering
      * the event loop since there is a busy Lua script running in timeout
      * condition, to avoid mixing the propagation of scripts with the
-     * propagation of DELs due to eviction. */
-    if (server.maxmemory && !scriptIsTimedout()) {
+     * propagation of DELs due to eviction.
+     *
+     * Also notice that if maxmemory is not configured, if server.script_oom is
+     * true we still need to verify that we are still out of memory, otherwise
+     * it could be that were out of memory and then maxmemory was set to unlimited.
+     * In such case we still want to enter this section to turn server.script_oom off
+     * as well. */
+    if ((server.maxmemory || server.script_oom) && !scriptIsTimedout()) {
```

Comment:
i don't understand this comment (maybe it needs to be cleared out a bit).
if it's about clearing server.script_oom, i think it could be an `else-if` for this `if`.
i.e. 
```
    if (server.maxmemory && !scriptIsTimedout()) {
        do eviction things.
        if (eval / evalsha / fcall)
            server.script_oom = out_of_memory;
    else if (!scriptIsTimedout()) {
        server.script_oom = false;
    }

Reply:
@oranagra I just set it to 0 anyway before the max memory check, worst case it will be set back to 1, let me know what you think.